app/assets/src/components/views/PipelineViz.jsx

```diff
@@ -0,0 +1,337 @@
+        yCenteredDOMPos = graph.getNodePosition(START_NODE_ID).y;
+      } else {
+        const xStartNodePos = graph.getNodePosition(START_NODE_ID).x;
+        graph.moveNodeToPosition(START_NODE_ID, xStartNodePos, yCenteredDOMPos);
+
+        const xEndNodePos = graph.getNodePosition(END_NODE_ID).x;
+        graph.moveNodeToPosition(END_NODE_ID, xEndNodePos, yCenteredDOMPos);
+      }
+    });
+  }
+
+  drawGraphs() {
+    this.modifyStepNames();
+    this.stageNames.forEach((stageName, i) => {
+      this.drawStageGraph(i);
+    });
+    this.adjustGraphNodePositions();
+    this.closeNonactiveSteps();
+  }
+
+  closeNonactiveSteps() {
+    this.stageNames.forEach((stageName, i) => {
+      const graph = this.graphs[i];
+      const stageData = this.props.stageResults.stages[stageName];
+      if (!(stageData.job_status == "STARTED")) {
+        graph.afterDrawingOnce(() => {
+          this.toggleStage(i);
+        });
+      }
+    });
+  }
+
+  drawStageGraph(index) {
+    const container = this[`container${index}`];
+
+    const nodeData = [];
+    const edgeData = [];
+
+    this.populateNodeAndEdgeData(index, nodeData, edgeData);
+
+    const options = {
+      nodes: {
+        borderWidth: 0,
+        color: this.props.nodeColor,
+        shape: "box",
+        shapeProperties: {
+          borderRadius: 6,
+        },
+        widthConstraint: {
+          minimum: 120,
+        },
+        heightConstraint: {
+          minimum: 24,
+        },
+        font: {
+          face: "Open Sans",
+        },
+      },
+      groups: {
+        startEndNodes: {
+          widthConstraint: 8,
+          heightConstraint: 0,
+          color: this.props.backgroundColor,
+          fixed: {
+            x: true,
+            y: true,
+          },
+        },
+      },
+      edges: {
+        arrows: {
+          to: {
+            enabled: true,
+            type: "arrow",
+            scaleFactor: 0.8,
+          },
+        },
+        smooth: {
+          type: "cubicBezier",
+          roundness: 0.8,
+        },
+        color: this.props.edgeColor,
+      },
+      layout: {
+        hierarchical: {
+          direction: "LR",
+          sortMethod: "directed",
+          levelSeparation: 200,
+          parentCentralization: false,
+          blockShifting: false,
+          edgeMinimization: false,
+        },
+      },
+      physics: {
+        enabled: false,
+      },
+      interaction: {
+        zoomView: false,
+        dragView: false,
+        dragNodes: false,
+      },
+    };
+
+    const currStageGraph = new NetworkGraph(
+      container,
+      nodeData,
+      edgeData,
+      options
+    );
+    currStageGraph.minimizeWidthGivenScale(1.0);
+
+    this.graphs.push(currStageGraph);
+  }
+
+  render() {
+    const stageContainers = [];
+
+    this.stageNames.forEach((stageName, i) => {
+      const isOpened = this.state.stagesOpened[i];
+
+      stageContainers.push(
+        <div key={i} className={cs.stage}>
```

Comment:
Not sure if the order of items in this array will ever change, but if so React recommends against using the index in a key: https://reactjs.org/docs/lists-and-keys.html#keys